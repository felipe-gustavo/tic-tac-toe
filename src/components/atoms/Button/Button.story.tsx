import { ComponentMeta, Story } from "@storybook/react";

import { Button, ButtonProps } from "./Button";

type InputPropOverrides = {
  text: string;
};

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: "Components/Button",
  component: Button,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    text: {
      control: "text",
      defaultValue: "button text",
    },
    type: {
      control: "select",
      options: ["primary", "secondary"],
      defaultValue: "primary",
    },
  },
} as ComponentMeta<typeof Button>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: Story<ButtonProps & InputPropOverrides> = ({
  // eslint-disable-next-line react/prop-types
  text,
  ...args
}) => <Button {...args}>{text}</Button>;

export const Primary = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Primary.args = {
  type: "primary",
};

export const Secondary = Template.bind({});
Secondary.args = {
  type: "secondary",
};
