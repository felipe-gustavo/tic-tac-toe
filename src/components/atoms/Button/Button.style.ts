import styled from "styled-components";

export const ButtonStyled = styled.button`
  padding: 12px 16px;
  text-transform: uppercase;
  font-size: 12px;
  font-weight: 600;
  border: 1px solid transparent;
  border-radius: 8px;
  transition: background-color 100ms ease-in-out;

  &.primary {
    border-color: #0a8ef3;
    background-color: #0a8ef3;
    color: white;

    &:hover,
    &:focus-visible,
    &:active {
      background-color: #6bbfff;
    }
  }

  &.secondary {
    border-color: #a7a7a7;
    background-color: #a7a7a7;
    color: white;

    &:hover,
    &:focus-visible,
    &:active {
      background-color: #d0d0d0;
    }
  }
`;
