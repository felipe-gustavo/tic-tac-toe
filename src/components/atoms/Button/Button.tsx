import clsx from "clsx";
import React from "react";

import { ButtonStyled } from "./Button.style";

export type ButtonProps = Omit<
  React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  >,
  "type" | "ref"
> & {
  type?: "primary" | "secondary";
};

export function Button({ type, className, ...props }: ButtonProps) {
  const buttonClasses = clsx([
    type,
    ...(className ? className.split(" ") : []),
  ]);

  return <ButtonStyled {...props} className={buttonClasses} type="button" />;
}

Button.defaultProps = {
  type: "primary",
};
