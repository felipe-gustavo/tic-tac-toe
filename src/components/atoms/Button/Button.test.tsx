import { render, screen } from "@testing-library/react";

import { Button, ButtonProps } from "./Button";

type ButtonType = NonNullable<ButtonProps["type"]>;

const renderComponent = (props: ButtonProps) =>
  render(<Button {...props}>some-button-text</Button>);

describe("<Button />", () => {
  it("is mixing the className", () => {
    renderComponent({ className: "my-custom-class" });

    const button = screen.getByRole("button");

    expect(button.classList).toContain("my-custom-class");
  });

  it.each<ButtonType>(["primary", "secondary"])(
    "match screenshot with %s type",
    (type: ButtonType) => {
      const { baseElement } = renderComponent({ type });

      expect(baseElement).toMatchSnapshot();
    }
  );

  it("has passing forward the rest of props", () => {
    renderComponent({ "data-testid": "some-testid" } as ButtonProps);

    const button = screen.queryByTestId("some-testid");

    expect(button).not.toBeUndefined();
  });
});
