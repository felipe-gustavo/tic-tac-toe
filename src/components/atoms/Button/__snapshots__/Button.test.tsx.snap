// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> match screenshot with primary type 1`] = `
<body>
  <div>
    <button
      class="sc-bczRLJ UDjBw primary"
      type="button"
    >
      some-button-text
    </button>
  </div>
</body>
`;

exports[`<Button /> match screenshot with secondary type 1`] = `
<body>
  <div>
    <button
      class="sc-bczRLJ UDjBw secondary"
      type="button"
    >
      some-button-text
    </button>
  </div>
</body>
`;
